File: apps/web/src/lib/evidence-bundle.ts
/** Exact UTF-8 logical-file verification; no uploads, extraction, or content rendering. */
export const MAX_BUNDLE_BYTES = 2 * 1024 * 1024;
const MAX_FILE_BYTES = 256 * 1024;
const MAX_NODES = 100_000;
const MAX_DEPTH = 32;
export const BUNDLE_LIMITATIONS = [
  "VALID means the supplied logical files match the supplied manifest's SHA-256 hashes and byte lengths.",
  "The manifest is unsigned and replaceable: replacing both content and its recorded hash cannot be detected here. Case, time, application, and projection metadata are not authenticated.",
  "Verification does not prove source telemetry truth, privileged-database integrity, authorship, or chain of custody.",
] as const;
const FIXED_PATHS = [
  "incident.json",
  "timeline.json",
  "alerts.json",
  "findings.json",
  "hypotheses.json",
  "notes.json",
  "audit.json",
  "entities.json",
  "README.txt",
];

export type BundleStatus =
  "VALID" | "MODIFIED" | "MISSING FILE" | "UNEXPECTED FILE" | "INVALID";
export type BundleFinding = {
  status: BundleStatus;
  path?: string;
  code: string;
};
export type BundleFileHash = { path: string; sha256: string; bytes: number };
export type BundleManifest = {
  format_version: "1";
  incident_id: string;
  scenario_id: "atlas-compromise" | null;
  projection: "public_synthetic" | "local_review";
  generated_at: string;
  evidence_ids: string[];
  files: BundleFileHash[];
  application: { name: "AegisGraph"; version: "0.1.0" };
  limitations: string[];
};
export type BundleVerification = {
  status: BundleStatus;
  findings: BundleFinding[];
  checked_files: number;
  limitations: string[];
  /** Schema-checked, still unsigned metadata; never includes logical file contents. */
  manifest?: BundleManifest;
};

class InvalidBundle extends Error {}
function invalid(code: string): never {
  throw new InvalidBundle(code);
}

// Retain integer-token identity: Python deliberately rejects bytes: 1.0 and 1e0.
class JsonNumber {
  constructor(
    readonly value: number,
    readonly integer: boolean,
  ) {}
}
type Json =
  null | boolean | string | JsonNumber | Json[] | { [key: string]: Json };
type JsonObject = { [key: string]: Json };

function utf8(value: string): Uint8Array<ArrayBuffer> {
  if (value.length > MAX_BUNDLE_BYTES) invalid("bundle_bytes_exceeded");
  for (let i = 0; i < value.length; i++) {
    const unit = value.charCodeAt(i);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(++i);
      if (!(next >= 0xdc00 && next <= 0xdfff)) invalid("invalid_utf8");
    } else if (unit >= 0xdc00 && unit <= 0xdfff) invalid("invalid_utf8");
  }
  return new TextEncoder().encode(value);
}

/** A byte-, depth-, and node-bounded JSON grammar, preserving duplicate keys. */
function parse(raw: string | Uint8Array): Json {
  let text: string;
  if (typeof raw === "string") text = raw;
  else if (
    ArrayBuffer.isView(raw) &&
    Object.prototype.toString.call(raw) === "[object Uint8Array]"
  ) {
    if (raw.byteLength > MAX_BUNDLE_BYTES) invalid("bundle_bytes_exceeded");
    try {
      text = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(
        raw,
      );
    } catch {
      invalid("invalid_utf8");
    }
  } else invalid("invalid_container");
  if (utf8(text).byteLength > MAX_BUNDLE_BYTES)
    invalid("bundle_bytes_exceeded");
  let position = 0,
    nodes = 0,
    stringBytes = 0;
  const number = /-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?/y;
  function node(depth: number) {
    if (++nodes > MAX_NODES || depth > MAX_DEPTH)
      invalid("structure_limit_exceeded");
  }
  function whitespace() {
    while (position < text.length && /[\x20\t\r\n]/.test(text[position]))
      position++;
  }
  function string(): string {
    const start = position++;
    while (position < text.length) {
      const char = text[position++];
      if (char === "\\") position++;
      else if (char === '"') {
        let value: string;
        try {
          value = JSON.parse(text.slice(start, position)) as string;
        } catch {
          invalid("invalid_json");
        }
        stringBytes += utf8(value).byteLength;
        if (stringBytes > MAX_BUNDLE_BYTES) invalid("bundle_bytes_exceeded");
        return value;
      }
    }
    invalid("invalid_json");
  }
  function value(depth: number): Json {
    node(depth);
    whitespace();
    const char = text[position];
    if (char === '"') return string();
    if (char === "{" || char === "[") {
      // Matches the Python pre-parser's maximum of 32 open containers.
      if (depth >= MAX_DEPTH) invalid("structure_limit_exceeded");
      const object = char === "{";
      position++;
      whitespace();
      const result: JsonObject | Json[] = object
        ? (Object.create(null) as JsonObject)
        : [];
      const end = object ? "}" : "]";
      if (text[position] === end) {
        position++;
        return result;
      }
      while (position < text.length) {
        whitespace();
        if (object) {
          if (text[position] !== '"') invalid("invalid_json");
          node(depth + 1);
          const key = string();
          whitespace();
          if (Object.hasOwn(result, key)) invalid("duplicate_json_key");
          if (text[position++] !== ":") invalid("invalid_json");
          (result as JsonObject)[key] = value(depth + 1);
        } else (result as Json[]).push(value(depth + 1));
        whitespace();
        if (text[position] === end) {
          position++;
          return result;
        }
        if (text[position++] !== ",") invalid("invalid_json");
      }
      invalid("invalid_json");
    }
    for (const [token, result] of [
      ["true", true],
      ["false", false],
      ["null", null],
    ] as const) {
      if (text.startsWith(token, position)) {
        position += token.length;
        return result;
      }
    }
    number.lastIndex = position;
    const match = number.exec(text);
    if (!match) invalid("invalid_json");
    position = number.lastIndex;
    const result = Number(match[0]);
    if (!Number.isFinite(result)) invalid("invalid_json_value");
    return new JsonNumber(result, !/[.eE]/.test(match[0]));
  }
  const result = value(0);
  whitespace();
  if (position !== text.length) invalid("invalid_json");
  return result;
}

function record(value: Json, fields: string[], code: string): JsonObject {
  if (
    value === null ||
    typeof value !== "object" ||
    Array.isArray(value) ||
    value instanceof JsonNumber
  )
    invalid(code);
  const keys = Object.keys(value);
  if (
    keys.length !== fields.length ||
    fields.some((field) => !Object.hasOwn(value, field))
  )
    invalid(code);
  return value;
}
function list(value: Json, maximum: number): Json[] {
  if (!Array.isArray(value) || value.length > maximum)
    invalid("collection_limit_exceeded");
  return value;
}
function matches(value: Json, pattern: RegExp): value is string {
  return typeof value === "string" && pattern.exec(value)?.[0] === value;
}
function identifier(value: Json, prefix: string): string {
  if (
    !matches(value, /^[A-Za-z][A-Za-z0-9-]{0,79}$/) ||
    !value.startsWith(prefix + "-") ||
    !/^[A-Za-z0-9]/.test(value.slice(prefix.length + 1))
  )
    invalid("invalid_identifier");
  return value;
}
function path(value: Json): string {
  if (
    !matches(value, /^[A-Za-z0-9][A-Za-z0-9._/-]{0,119}$/) ||
    value
      .split("/")
      .some(
        (part) =>
          !part ||
          part === "." ||
          part === ".." ||
          part.endsWith(".") ||
          /^(?:con|prn|aux|nul|com[1-9]|lpt[1-9])(?:\.|$)/i.test(part),
      )
  )
    invalid("unsafe_path");
  return value;
}
function timestamp(value: Json): string {
  if (typeof value !== "string") invalid("invalid_timestamp");
  const match =
    /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{6}))?\+00:00$/.exec(
      value,
    );
  if (!match || match[0] !== value) invalid("invalid_timestamp");
  const [, year, month, day, hour, minute, second, micro] = match;
  const y = Number(year),
    m = Number(month),
    d = Number(day);
  const leap = y % 4 === 0 && (y % 100 !== 0 || y % 400 === 0);
  const days = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (
    y < 1 ||
    m < 1 ||
    m > 12 ||
    d < 1 ||
    d > days[m - 1] ||
    Number(hour) > 23 ||
    Number(minute) > 59 ||
    Number(second) > 59 ||
    micro === "000000"
  )
    invalid("invalid_timestamp");
  return value;
}
function uniquePaths(rows: Json[], manifest: boolean): Map<string, JsonObject> {
  const result = new Map<string, JsonObject>();
  const folded = new Set<string>();
  for (const item of rows) {
    const row = record(
      item,
      manifest ? ["path", "sha256", "bytes"] : ["path", "content"],
      "invalid_file_entry",
    );
    const name = path(row.path);
    if (folded.has(name.toLowerCase())) invalid("duplicate_path");
    folded.add(name.toLowerCase());
    if (manifest) {
      if (!matches(row.sha256, /^[a-f0-9]{64}$/)) invalid("invalid_hash");
      const size = row.bytes;
      if (
        !(size instanceof JsonNumber) ||
        !size.integer ||
        !Number.isSafeInteger(size.value) ||
        size.value < 0 ||
        size.value > MAX_FILE_BYTES
      )
        invalid("invalid_file_size");
    } else if (
      typeof row.content !== "string" ||
      utf8(row.content).byteLength > MAX_FILE_BYTES
    )
      invalid("file_bytes_exceeded");
    result.set(name, row);
  }
  return result;
}
function container(raw: string | Uint8Array) {
  const value = record(
    parse(raw),
    ["format_version", "manifest", "files"],
    "invalid_container",
  );
  if (value.format_version !== "1") invalid("invalid_container");
  const manifest = record(
    value.manifest,
    [
      "format_version",
      "incident_id",
      "scenario_id",
      "projection",
      "generated_at",
      "evidence_ids",
      "files",
      "application",
      "limitations",
    ],
    "invalid_manifest",
  );
  if (manifest.format_version !== "1") invalid("invalid_manifest");
  const incident = identifier(manifest.incident_id, "INC");
  const generated = timestamp(manifest.generated_at);
  if (
    (manifest.scenario_id !== null &&
      manifest.scenario_id !== "atlas-compromise") ||
    (manifest.projection !== "public_synthetic" &&
      manifest.projection !== "local_review")
  )
    invalid("invalid_manifest");
  const application = record(
    manifest.application,
    ["name", "version"],
    "invalid_manifest",
  );
  const limitations = manifest.limitations;
  if (
    application.name !== "AegisGraph" ||
    application.version !== "0.1.0" ||
    !Array.isArray(limitations) ||
    limitations.length !== BUNDLE_LIMITATIONS.length ||
    BUNDLE_LIMITATIONS.some((line, index) => limitations[index] !== line)
  )
    invalid("invalid_manifest");
  const ids = list(manifest.evidence_ids, 80).map((item) =>
    identifier(item, "EVD"),
  );
  if (new Set(ids).size !== ids.length) invalid("duplicate_identifier");
  const expected = uniquePaths(list(manifest.files, 100), true);
  const actual = uniquePaths(list(value.files, 100), false);
  const inventory = new Set([
    ...FIXED_PATHS,
    ...ids.map((id) => `evidence/${id}.json`),
  ]);
  if (
    expected.size !== inventory.size ||
    [...expected.keys()].some((name) => !inventory.has(name))
  )
    invalid("invalid_manifest_inventory");
  const folded = new Map(
    [...expected.keys()].map((name) => [name.toLowerCase(), name]),
  );
  if (
    [...actual.keys()].some(
      (name) =>
        folded.has(name.toLowerCase()) &&
        name !== folded.get(name.toLowerCase()),
    )
  )
    invalid("duplicate_path");
  const metadata: BundleManifest = {
    format_version: "1",
    incident_id: incident,
    scenario_id: manifest.scenario_id,
    projection: manifest.projection,
    generated_at: generated,
    evidence_ids: ids,
    application: { name: "AegisGraph", version: "0.1.0" },
    limitations: [...BUNDLE_LIMITATIONS],
    files: [...expected].map(([name, row]) => ({
      path: name,
      sha256: row.sha256 as string,
      bytes: (row.bytes as JsonNumber).value,
    })),
  };
  return { expected, actual, metadata };
}

export async function verifyBundle(
  raw: string | Uint8Array,
): Promise<BundleVerification> {
  let bundle: ReturnType<typeof container>;
  try {
    bundle = container(raw);
  } catch (error) {
    if (!(error instanceof InvalidBundle)) throw error;
    return {
      status: "INVALID",
      findings: [{ status: "INVALID", code: error.message }],
      checked_files: 0,
      limitations: [...BUNDLE_LIMITATIONS],
    };
  }
  if (!globalThis.crypto?.subtle)
    throw new Error("Secure browser SHA-256 verification is unavailable.");
  const findings: BundleFinding[] = [];
  let checked = 0;
  for (const [name, entry] of [...bundle.expected].sort(([left], [right]) =>
    left < right ? -1 : left > right ? 1 : 0,
  )) {
    const file = bundle.actual.get(name);
    if (!file) {
      findings.push({
        status: "MISSING FILE",
        path: name,
        code: "file_missing",
      });
      continue;
    }
    checked++;
    const bytes = utf8(file.content as string);
    let digest: ArrayBuffer;
    try {
      digest = await globalThis.crypto.subtle.digest("SHA-256", bytes);
    } catch {
      throw new Error("Secure browser SHA-256 verification is unavailable.");
    }
    const hash = Array.from(new Uint8Array(digest), (byte) =>
      byte.toString(16).padStart(2, "0"),
    ).join("");
    if (
      bytes.byteLength !== (entry.bytes as JsonNumber).value ||
      hash !== entry.sha256
    )
      findings.push({
        status: "MODIFIED",
        path: name,
        code: "content_mismatch",
      });
  }
  for (const name of [...bundle.actual.keys()]
    .filter((name) => !bundle.expected.has(name))
    .sort())
    findings.push({
      status: "UNEXPECTED FILE",
      path: name,
      code: "file_unlisted",
    });
  const status =
    (["MODIFIED", "MISSING FILE", "UNEXPECTED FILE"] as const).find(
      (candidate) => findings.some((finding) => finding.status === candidate),
    ) ?? "VALID";
  return {
    status,
    findings,
    checked_files: checked,
    limitations: [...BUNDLE_LIMITATIONS],
    manifest: bundle.metadata,
  };
}

File: apps/api/aegisgraph/evidence_bundle.py
"""Bounded logical-file exports and exact UTF-8 SHA-256 verification.

No archives are extracted and no file paths are opened. A valid result establishes
agreement with the supplied unsigned manifest, not provenance or source truth.
"""

from __future__ import annotations

import hashlib
import json
import math
import re
from datetime import UTC, datetime

from pydantic import ValidationError

from .schema import CanonicalEvent

FORMAT_VERSION = "1"
MAX_BUNDLE_BYTES = 2 * 1024 * 1024
MAX_FILE_BYTES = 256 * 1024
MAX_FILES = 100
MAX_EVIDENCE = 80
MAX_DEPTH = 32
MAX_NODES = 100_000
LIMITATIONS = (
    "VALID means the supplied logical files match the supplied manifest's SHA-256 hashes and byte lengths.",
    "The manifest is unsigned and replaceable: replacing both content and its recorded hash cannot be detected here. Case, time, application, and projection metadata are not authenticated.",
    "Verification does not prove source telemetry truth, privileged-database integrity, authorship, or chain of custody.",
)
FIXED_PATHS = frozenset(
    {
        "incident.json",
        "timeline.json",
        "alerts.json",
        "findings.json",
        "hypotheses.json",
        "notes.json",
        "audit.json",
        "entities.json",
        "README.txt",
    }
)
_SNAPSHOT_FIELDS = {
    "incident",
    "scenario_id",
    "projection",
    "evidence",
    "alerts",
    "findings",
    "hypotheses",
    "notes",
    "audit",
    "entities",
}
_MANIFEST_FIELDS = {
    "format_version",
    "incident_id",
    "scenario_id",
    "projection",
    "generated_at",
    "evidence_ids",
    "files",
    "application",
    "limitations",
}
_ID = re.compile(r"^[A-Za-z][A-Za-z0-9-]{0,79}$")
_PATH = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._/-]{0,119}$")
_HASH = re.compile(r"^[a-f0-9]{64}$")
_RESERVED = re.compile(r"^(?:con|prn|aux|nul|com[1-9]|lpt[1-9])(?:\.|$)", re.IGNORECASE)


class BundleInputError(ValueError):
    """Fixed safe diagnostics, without caller-controlled content or identifiers."""


def _fail(code):
    raise BundleInputError(code)


def _json_shape(value):
    """Bound nested objects before copying, encoding, or hashing them."""
    nodes = 0
    text_bytes = 0
    stack = [(value, 0)]
    while stack:
        item, depth = stack.pop()
        nodes += 1
        if depth > MAX_DEPTH or nodes > MAX_NODES:
            _fail("structure_limit_exceeded")
        if type(item) is dict:
            if len(item) * 2 + len(stack) + nodes > MAX_NODES:
                _fail("structure_limit_exceeded")
            for key, child in item.items():
                if type(key) is not str:
                    _fail("invalid_json_value")
                stack.append((key, depth + 1))
                stack.append((child, depth + 1))
        elif type(item) is list:
            if len(item) + len(stack) + nodes > MAX_NODES:
                _fail("structure_limit_exceeded")
            stack.extend((child, depth + 1) for child in item)
        elif type(item) is str:
            if len(item) > MAX_BUNDLE_BYTES:
                _fail("bundle_bytes_exceeded")
            try:
                text_bytes += len(item.encode("utf-8"))
            except UnicodeError as exc:
                raise BundleInputError("invalid_utf8") from exc
            if text_bytes > MAX_BUNDLE_BYTES:
                _fail("bundle_bytes_exceeded")
        elif type(item) is float:
            if not math.isfinite(item):
                _fail("invalid_json_value")
        elif item is not None and type(item) not in (bool, int):
            _fail("invalid_json_value")


def _encode(value):
    _json_shape(value)
    try:
        encoded = json.dumps(
            value, sort_keys=True, separators=(",", ":"), ensure_ascii=False, allow_nan=False
        ).encode("utf-8")
    except (TypeError, ValueError, RecursionError, UnicodeError) as exc:
        raise BundleInputError("invalid_json_value") from exc
    if len(encoded) > MAX_BUNDLE_BYTES:
        _fail("bundle_bytes_exceeded")
    return encoded


def _pairs(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            _fail("duplicate_json_key")
        result[key] = value
    return result


def _constant(_value):
    _fail("invalid_json_value")


def _check_text_depth(text):
    """Reject deeply nested JSON before the recursive standard-library parser."""
    depth = 0
    quoted = escaped = False
    for char in text:
        if quoted:
            if escaped:
                escaped = False
            elif char == "\\":
                escaped = True
            elif char == '"':
                quoted = False
        elif char == '"':
            quoted = True
        elif char in "[{":
            depth += 1
            if depth > MAX_DEPTH:
                _fail("structure_limit_exceeded")
        elif char in "]}":
            depth -= 1


def _load(value):
    if type(value) is dict:
        return json.loads(_encode(value), object_pairs_hook=_pairs)
    if type(value) is bytes:
        if len(value) > MAX_BUNDLE_BYTES:
            _fail("bundle_bytes_exceeded")
        try:
            value = value.decode("utf-8")
        except UnicodeError as exc:
            raise BundleInputError("invalid_utf8") from exc
    if type(value) is not str:
        _fail("invalid_container")
    try:
        if len(value.encode("utf-8")) > MAX_BUNDLE_BYTES:
            _fail("bundle_bytes_exceeded")
    except UnicodeError as exc:
        raise BundleInputError("invalid_utf8") from exc
    _check_text_depth(value)
    try:
        result = json.loads(value, object_pairs_hook=_pairs, parse_constant=_constant)
    except (ValueError, RecursionError) as exc:
        if isinstance(exc, BundleInputError):
            raise
        raise BundleInputError("invalid_json") from exc
    _json_shape(result)
    return result


def _identifier(value, prefix):
    if (
        type(value) is not str
        or not _ID.fullmatch(value)
        or not value.startswith(prefix + "-")
        or len(value) <= len(prefix) + 1
        or not value[len(prefix) + 1].isalnum()
    ):
        _fail("invalid_identifier")
    return value


def _timestamp(value):
    if type(value) is not str or len(value) > 40:
        _fail("invalid_timestamp")
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            raise ValueError("timezone required")
        return parsed.astimezone(UTC)
    except (ValueError, OverflowError) as exc:
        raise BundleInputError("invalid_timestamp") from exc


def _list(value, maximum):
    if type(value) is not list or len(value) > maximum:
        _fail("collection_limit_exceeded")
    return value


def _references(value, allowed, maximum=MAX_EVIDENCE):
    rows = _list(value, maximum)
    if any(type(item) is not str for item in rows) or len(set(rows)) != len(rows):
        _fail("invalid_reference")
    if not set(rows) <= allowed:
        _fail("cross_scope_reference")


def _scoped_rows(value, incident_id, prefix, maximum):
    rows = _list(value, maximum)
    seen = set()
    for row in rows:
        if type(row) is not dict or row.get("incident_id") != incident_id:
            _fail("cross_scope_row")
        identifier = _identifier(row.get("id"), prefix)
        if identifier in seen:
            _fail("duplicate_identifier")
        seen.add(identifier)
    return seen


def _audit_references(rows, scope, evidence_ids, event_ids, finding_ids, note_ids, hypothesis_ids):
    """Check structured provenance references; never interpret arbitrary human prose."""
    references = {
        "evidence_ids": evidence_ids,
        "supporting_evidence_ids": evidence_ids,
        "contradicting_evidence_ids": evidence_ids,
        "event_ids": event_ids,
        "finding_ids": finding_ids,
        "related_finding_ids": finding_ids,
    }
    single_references = {
        "evidence_id": evidence_ids,
        "event_id": event_ids,
        "finding_id": finding_ids,
        "hypothesis_id": hypothesis_ids,
    }
    objects = {
        "incident_created": {scope},
        "status_changed": {scope},
        "severity_changed": {scope},
        "analyst_assigned": {scope},
        "ai_analysis_requested": {scope},
        "evidence_annotated": evidence_ids,
        "finding_created": finding_ids,
        "finding_reviewed": finding_ids,
        "note_created": note_ids,
    }
    for row in rows:
        action = row.get("action")
        if type(action) is not str:
            _fail("invalid_audit_action")
        if action in objects and (
            type(row.get("object_id")) is not str or row["object_id"] not in objects[action]
        ):
            _fail("cross_scope_audit_object")
        if action == "hypothesis_reviewed":
            # Revision membership is checked by the scoped database snapshot.
            # The audit preserves its real revision object, not a substituted HYP ID.
            _identifier(row.get("object_id"), "HPR")
            after = row.get("after")
            if type(after) is not dict:
                _fail("invalid_audit_revision")
            _references([after.get("hypothesis_id")], hypothesis_ids)
        stack = [row.get("before"), row.get("after")]
        while stack:
            value = stack.pop()
            if type(value) is dict:
                for key, child in value.items():
                    if key in references:
                        _references(child, references[key])
                    elif key in single_references:
                        _references([child], single_references[key])
                    elif key in ("incident_id", "scope_id") and child != scope:
                        _fail("cross_scope_audit_reference")
                    stack.append(child)
            elif type(value) is list:
                stack.extend(value)


def _validate_snapshot(snapshot):
    if type(snapshot) is not dict or set(snapshot) != _SNAPSHOT_FIELDS:
        _fail("invalid_snapshot")
    if snapshot["projection"] not in ("public_synthetic", "local_review"):
        _fail("invalid_projection")
    if snapshot["scenario_id"] not in (None, "atlas-compromise"):
        _fail("invalid_scenario")
    if type(snapshot["incident"]) is not dict:
        _fail("invalid_incident")
    scope = _identifier(snapshot["incident"].get("id"), "INC")
    evidence = snapshot["evidence"]
    evidence_ids = _scoped_rows(evidence, scope, "EVD", MAX_EVIDENCE)
    event_ids = set()
    for row in evidence:
        event_id = _identifier(row.get("event_id"), "EVT")
        if event_id in event_ids:
            _fail("duplicate_event")
        event_ids.add(event_id)
        event = row.get("event")
        if type(event) is not dict or event.get("event_id") != event_id:
            _fail("event_reference_mismatch")
        if _timestamp(row.get("timestamp")) != _timestamp(event.get("timestamp")):
            _fail("event_timestamp_mismatch")
        try:
            CanonicalEvent.model_validate(event)
        except (ValidationError, ValueError, TypeError) as exc:
            raise BundleInputError("invalid_canonical_event") from exc
    _scoped_rows(snapshot["alerts"], scope, "ALT", 80)
    for row in snapshot["alerts"]:
        _references(row.get("event_ids"), event_ids)
    finding_ids = _scoped_rows(snapshot["findings"], scope, "FND", 50)
    for row in snapshot["findings"]:
        _references(row.get("evidence_ids"), evidence_ids)
    note_ids = _scoped_rows(snapshot["notes"], scope, "NOTE", 100)
    _scoped_rows(snapshot["audit"], scope, "AUD", 1000)
    ledger = snapshot["hypotheses"]
    if type(ledger) is not dict or ledger.get("scope_id") != scope:
        _fail("cross_scope_hypothesis")
    seen_hypotheses = set()
    for row in _list(ledger.get("items"), 4):
        if type(row) is not dict:
            _fail("invalid_hypothesis")
        identifier = _identifier(row.get("id"), "HYP")
        if identifier in seen_hypotheses:
            _fail("duplicate_identifier")
        seen_hypotheses.add(identifier)
        provenance = row.get("provenance")
        if type(provenance) is not dict or provenance.get("scope_id") != scope:
            _fail("cross_scope_hypothesis")
        for field in ("supporting_evidence_ids", "contradicting_evidence_ids"):
            _references(row.get(field), evidence_ids)
        _references(row.get("related_finding_ids"), finding_ids, 50)
        if "review" in row:
            if type(row["review"]) is not dict:
                _fail("invalid_review")
            _references(row["review"].get("related_finding_ids"), finding_ids, 50)
    _audit_references(
        snapshot["audit"], scope, evidence_ids, event_ids, finding_ids, note_ids, seen_hypotheses
    )
    entities = snapshot["entities"]
    if type(entities) is not dict or set(entities) != {"nodes", "edges"}:
        _fail("invalid_entities")
    nodes = _list(entities["nodes"], 720)
    node_ids = set()
    for row in nodes:
        if type(row) is not dict:
            _fail("invalid_entities")
        identifier = _identifier(row.get("id"), "ENT")
        if identifier in node_ids:
            _fail("duplicate_identifier")
        node_ids.add(identifier)
    for row in _list(entities["edges"], 640):
        if (
            type(row) is not dict
            or type(row.get("source")) is not str
            or type(row.get("target")) is not str
            or row.get("source") not in node_ids
            or row.get("target") not in node_ids
        ):
            _fail("invalid_entity_reference")
        _references(row.get("evidence_ids"), evidence_ids)
    if snapshot["projection"] == "public_synthetic" and (
        snapshot["incident"].get("owner") is not None
        or any(row.get("note") not in (None, "") for row in evidence)
        or snapshot["findings"]
        or snapshot["notes"]
        or snapshot["audit"]
        or "context_digest" in ledger
        or any("review" in row for row in ledger["items"])
    ):
        _fail("private_public_projection")
    return scope, evidence_ids


def _readme(projection):
    return (
        "AegisGraph hash-verifiable synthetic investigation export\n\n"
        "This JSON container stores logical files as exact UTF-8 content strings. "
        "No paths need extraction. Hash the content string's UTF-8 bytes without "
        "normalizing Unicode or newlines. All telemetry is synthetic demo material.\n\n"
        f"Projection: {projection}. Public projections omit owner, evidence annotations, "
        "findings, notes, audit entries, and human hypothesis review state. Empty public "
        "collections describe that projection, not an assertion that no local actions exist. "
        "Local projections contain bounded committed case material; human actor labels "
        "are not authenticated identities.\n\n" + "\n".join(LIMITATIONS) + "\n"
    )


def build_bundle(snapshot: dict, *, generated_at: datetime) -> dict:
    """Build deterministic contents for one scoped snapshot and explicit export time."""
    snapshot = _load(snapshot)
    scope, evidence_ids = _validate_snapshot(snapshot)
    if not isinstance(generated_at, datetime) or generated_at.tzinfo is None:
        _fail("invalid_timestamp")
    try:
        timestamp = generated_at.astimezone(UTC).isoformat()
    except (ValueError, OverflowError) as exc:
        raise BundleInputError("invalid_timestamp") from exc
    evidence = sorted(
        snapshot["evidence"], key=lambda row: (_timestamp(row["timestamp"]), row["id"])
    )
    logical = {
        "incident.json": snapshot["incident"],
        "timeline.json": [
            {key: row[key] for key in ("id", "event_id", "timestamp")} for row in evidence
        ],
        "hypotheses.json": snapshot["hypotheses"],
        "entities.json": {
            "nodes": sorted(snapshot["entities"]["nodes"], key=lambda row: row["id"]),
            "edges": sorted(
                snapshot["entities"]["edges"],
                key=lambda row: (row["source"], row["target"], row.get("label", "")),
            ),
        },
    }
    for key in ("alerts", "findings", "notes", "audit"):
        logical[f"{key}.json"] = sorted(snapshot[key], key=lambda row: row["id"])
    for row in evidence:
        logical[f"evidence/{row['id']}.json"] = row
    contents = {path: _encode(value).decode("utf-8") + "\n" for path, value in logical.items()}
    contents["README.txt"] = _readme(snapshot["projection"])
    files, entries = [], []
    for path, content in sorted(contents.items()):
        encoded = content.encode("utf-8")
        if len(encoded) > MAX_FILE_BYTES:
            _fail("file_bytes_exceeded")
        files.append({"path": path, "content": content})
        entries.append(
            {"path": path, "sha256": hashlib.sha256(encoded).hexdigest(), "bytes": len(encoded)}
        )
    result = {
        "format_version": FORMAT_VERSION,
        "manifest": {
            "format_version": FORMAT_VERSION,
            "incident_id": scope,
            "scenario_id": snapshot["scenario_id"],
            "projection": snapshot["projection"],
            "generated_at": timestamp,
            "evidence_ids": sorted(evidence_ids),
            "files": entries,
            "application": {"name": "AegisGraph", "version": "0.1.0"},
            "limitations": list(LIMITATIONS),
        },
        "files": files,
    }
    _encode(result)
    _validate_container(result)
    return result


def _path(value):
    if type(value) is not str or not _PATH.fullmatch(value):
        _fail("unsafe_path")
    if any(
        part in ("", ".", "..") or part.endswith(".") or _RESERVED.match(part)
        for part in value.split("/")
    ):
        _fail("unsafe_path")
    return value


def _file_index(value, manifest):
    result, folded = {}, set()
    for row in _list(value, MAX_FILES):
        fields = {"path", "sha256", "bytes"} if manifest else {"path", "content"}
        if type(row) is not dict or set(row) != fields:
            _fail("invalid_file_entry")
        path = _path(row["path"])
        if path.casefold() in folded:
            _fail("duplicate_path")
        folded.add(path.casefold())
        if manifest:
            if type(row["sha256"]) is not str or not _HASH.fullmatch(row["sha256"]):
                _fail("invalid_hash")
            if type(row["bytes"]) is not int or not 0 <= row["bytes"] <= MAX_FILE_BYTES:
                _fail("invalid_file_size")
        elif (
            type(row["content"]) is not str or len(row["content"].encode("utf-8")) > MAX_FILE_BYTES
        ):
            _fail("file_bytes_exceeded")
        result[path] = row
    return result


def _validate_container(value):
    if (
        type(value) is not dict
        or set(value) != {"format_version", "manifest", "files"}
        or value["format_version"] != FORMAT_VERSION
    ):
        _fail("invalid_container")
    manifest = value["manifest"]
    if (
        type(manifest) is not dict
        or set(manifest) != _MANIFEST_FIELDS
        or manifest["format_version"] != FORMAT_VERSION
    ):
        _fail("invalid_manifest")
    _identifier(manifest["incident_id"], "INC")
    if _timestamp(manifest["generated_at"]).isoformat() != manifest["generated_at"]:
        _fail("invalid_timestamp")
    if manifest["scenario_id"] not in (None, "atlas-compromise") or manifest["projection"] not in (
        "public_synthetic",
        "local_review",
    ):
        _fail("invalid_manifest")
    if manifest["application"] != {"name": "AegisGraph", "version": "0.1.0"} or manifest[
        "limitations"
    ] != list(LIMITATIONS):
        _fail("invalid_manifest")
    ids = [_identifier(item, "EVD") for item in _list(manifest["evidence_ids"], MAX_EVIDENCE)]
    if len(set(ids)) != len(ids):
        _fail("duplicate_identifier")
    expected = _file_index(manifest["files"], True)
    actual = _file_index(value["files"], False)
    if set(expected) != FIXED_PATHS | {f"evidence/{identifier}.json" for identifier in ids}:
        _fail("invalid_manifest_inventory")
    expected_folded = {path.casefold(): path for path in expected}
    if any(
        path.casefold() in expected_folded and path != expected_folded[path.casefold()]
        for path in actual
    ):
        _fail("duplicate_path")
    return expected, actual


def verify_bundle(value: dict | str | bytes) -> dict:
    """Verify supplied exact strings; return safe findings without rendering contents."""
    try:
        expected, actual = _validate_container(_load(value))
    except BundleInputError as exc:
        return {
            "status": "INVALID",
            "findings": [{"status": "INVALID", "code": str(exc)}],
            "checked_files": 0,
            "limitations": list(LIMITATIONS),
        }
    findings = []
    checked = 0
    for path, entry in sorted(expected.items()):
        if path not in actual:
            findings.append({"status": "MISSING FILE", "path": path, "code": "file_missing"})
            continue
        checked += 1
        encoded = actual[path]["content"].encode("utf-8")
        if len(encoded) != entry["bytes"] or hashlib.sha256(encoded).hexdigest() != entry["sha256"]:
            findings.append({"status": "MODIFIED", "path": path, "code": "content_mismatch"})
    findings.extend(
        {"status": "UNEXPECTED FILE", "path": path, "code": "file_unlisted"}
        for path in sorted(set(actual) - set(expected))
    )
    statuses = {row["status"] for row in findings}
    status = next(
        (item for item in ("MODIFIED", "MISSING FILE", "UNEXPECTED FILE") if item in statuses),
        "VALID",
    )
    return {
        "status": status,
        "findings": findings,
        "checked_files": checked,
        "limitations": list(LIMITATIONS),
    }
